### PR TITLE
Understanding @Input and @Output Decorators in Angular

### LINKED ISSUE #2435: Understanding @Output and EventEmitter in Angular
### Proposed title of article
Understanding @Output and EventEmitter in Angular

### Introduction paragraph (2-3 paragraphs):

Angular is based on a one-directional data flow and does not have two-way data binding. So, how do you get a component to emit an event to another component?
In Angular, a component can emit an event using @Output and EventEmitter. Both are parts of the @angular/core.

Angular components are the fundamental building blocks of architecture. Once you’ve understood component architecture, you can build a picture in your head as to how to assemble your interface based on data communication.

In this post we’re going to dive into Angular components and how to use the EventEmitter and Outputs, so we can transfer or notify any parent component that something has changed, or we would like to change it. This is typically done via “events”, hence our “EventEmitter” and is designed around a uni-directional data flow system that adopts a much more reasonable approach to application development.

### Key takeaways: 
1.Understanding @Input decorator  
2.Understanding @Output decorator  
3. Angular Event Emitters

### PR DESCRIPTION
**Pull request by odiwuoramos**
### Article checklist

**ATTENTION: In your PR - add the ISSUE (topic form) # that this PR corresponds to. You can do this by also adding a comment that states **["This fixes #(enter your Idea Content Form #)"](https://youtu.be/TKJ4RdhyB5Y?t=278)** to link and close your Idea Content (issue) when the PR is merged.**
      
      
**NOTE: (Please ensure that you have only one open issue + linked pull request at a time. This will ensure that we complete the article in a timely manner from inception to publishing.)**

**If you have not already please go over our [Resources Page](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md) for more tips and tricks on how to improve our overall technical writing.**

#### Preliminary Checks - Formatting and Structure 
- [x] Does your article follow any of the suggested structure formats? - **see [example formats](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md#example-format-structure)**
- [x] Is your article properly formatted in Markdown? - **see [Github Markdown guide](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)**
- [x] Have you used the correct folder and file structure? - **see [Contribution Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**
- [x] Is your article over 750 words? - **Tool**: [Wordcounter.net](https://wordcounter.net)
- [x] Have you used our preferred heading structure? - **[H3 (###) and up](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers)**
- [x] Does your article provide enough value and detail about your topic? - **Articles should be clear, accurate, and fully explained.**
- [x] Can your article be understood by beginners? - **Assume the audience is smart but has no prior exposure to the common terminology in your article.**
- [x] Have you included a hero image that is 600x400 pixels and have the copyright to use it? - **Tools:** [Pixlr Image Editor](https://pixlr.com/e) for resizing and [Unsplash](https://unsplash.com) for Creative Commons images
- [x] In your PR - add a comment that states **["This fixes #(enter your Idea Content Form #)"](https://youtu.be/TKJ4RdhyB5Y?t=278)** to link and close your Idea Content (issue) when the PR is merged.
      
#### Grammar & plagiarism checks
- [x] Have you spellchecked and grammar checked your article? - **Tools:** [Write&Improve](https://writeandimprove.com/?lang=en-GB) - [ProWritingAid](https://prowritingaid.com/) -  [Grammarly](https://grammarly.com) 
- [x] Please place your article through a 3rd party plagiarism checker? We suggest using Quetext, [this tool](https://www.quetext.com/) is free and has a daily limit. **We typically accept articles with 10% or less.**
- [x] Have you checked your article for readability? - **Tool:** [Hemmingway](http://www.hemingwayapp.com)
- [x] Have you added sources for quotes and images that aren't yours?

#### Technical checks
- [x] Are your code snippets properly formatted for syntax highlighting - **see [Syntax guide](https://gohugo.io/content-management/syntax-highlighting)**

### Contribution guidelines
For first-time contributors and for more details, **see [Contributing Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**

This resolves #2435

### CODE REVIEW

**Comment by ahmadmardeni1** [comment]
@adrianmurage 👋
Please if you have time, review this one.

**Comment by odiwuoramos** [comment]
@ahmadmardeni1 , @louisefindlay23 , is the reviewer really free to look at my article?

**Comment by adrianmurage** [comment]
> 
> 
> @ahmadmardeni1 , @louisefindlay23 , is @adrianmurage really free to look at my article?

@odiwuoramos thank you for your patience.
I'm looking into it in the next few hours

File: content/articles/angular-output-and-input-decorator/index.md [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+<app-music-details [musicianName]="name"></app-music-details>
+
+```
+
+In the above example, we pass the value of the `musicianName` using the parent `name` property which we defined in the parent component and assigned the value to.  
+
+Serve the application by running the following command:  
+
+```bash
+cd output-decorator
```
Comment (adrianmurage) [inline content/articles/angular-output-and-input-decorator/index.md:128]: I do not see the relevance of this command as there is no directory named `output-decorator`
Consider revising it.

**Review by adrianmurage — CHANGES_REQUESTED** [review]
@odiwuoramos 
The app throws the following error when run:

``` bash
Error: src/app/music-details/music-details.component.ts:11:44 - error TS2558: Expected 0 type arguments, but got 1.

11   @Output() newMusician = new EventEmitter<string>();
```

Kindly take a look and @ me when you've fixed it

**Comment by odiwuoramos** [comment]
Hi, @adrianmurage , I've  fixed the code, could you re-execute the code again,,,you many as well have a look at the screeshots.
Thanks.

**Comment by hectorkambow** [comment]
plagiarism check done

File: content/articles/angular-output-and-input-decorator/index.md [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+
+![Input decorator output](/engineering-education/angular-output-and-input-decorator/input.png)
+
+### How does an Angular @Output work?
+
+In the previous section, we explored the usage of the child component and how to use the `@Input` decorator. In this section, we look at the `@Output` decorator, which is the reverse of the `@Input` decorator.
+
+This decorator allows for the sharing of data from the child component. It invokes an event that in turn notifies the parent component.  
+
+> It's important to note that the `@Output` decorator uses `EventEmitter` to accomplish its task.
```
Comment (adrianmurage) [inline content/articles/angular-output-and-input-decorator/index.md:135]: I've discovered that `EventEmitter` can be imported from multiple places. For instance, my import in the first review was:
```javascript
import { EventEmitter } from 'stream';
```
Which caused the error I was running into.

Therefore it's important to point out that `EventEmitter` should be imported from `@angular/core` like so:
```javascript
import { Component, Input, OnInit, Output, EventEmitter} from '@angular/core';
```

**Review by adrianmurage — CHANGES_REQUESTED** [review]
@odiwuoramos 
Great work!

I was able to figure out what was causing the said error. 
Kindly make the addition to the article to prevent the readers from facing the same challenges.

**Comment by odiwuoramos** [comment]
@adrianmurage, I've updated the EventEmitter imports

**Comment by adrianmurage** [comment]
@odiwuoramos 
Great work with this!
Looking forward to seeing more of your work

**Comment by ahmadmardeni1** [comment]
Good work 🚀 @odiwuoramos @adrianmurage 

Please work on the front matter, it should be complete before you put the "ready for final review" label.

**Pull request merged**